packages/grid/_modules_/grid/hooks/features/treeData/useGridTreeData.ts

```diff
@@ -0,0 +1,149 @@
+import * as React from 'react';
+import { GridApiRef } from '../../../models/api/gridApiRef';
+import { GridComponentProps } from '../../../GridComponentProps';
+import { GRID_TREE_DATA_GROUP_COL_DEF } from './gridTreeDataGroupColDef';
+import { useGridApiEventHandler } from '../../utils/useGridApiEventHandler';
+import { GridEvents } from '../../../constants';
+import {
+  GridCellParams,
+  GridColDef,
+  GridColDefOverrideParams,
+  GridColumns,
+  MuiEvent,
+} from '../../../models';
+import { isSpaceKey } from '../../../utils/keyboardUtils';
+import { useFirstRender } from '../../utils/useFirstRender';
+import { buildRowTree } from '../../../utils/rowTreeUtils';
+import { GridRowGroupingPreProcessing } from '../../core/rowGroupsPerProcessing';
+import { isFunction } from '../../../utils/utils';
+import { gridFilteredDescendantCountLookupSelector } from '../filter';
+import { GridPreProcessingGroup } from '../../core/preProcessing';
+
+/**
+ * Only available in DataGridPro
+ * @requires useGridPreProcessing (method)
+ * @requires useGridRowGroupsPreProcessing (method)
+ */
+export const useGridTreeData = (
+  apiRef: GridApiRef,
+  props: Pick<
+    GridComponentProps,
+    'treeData' | 'getTreeDataPath' | 'groupingColDef' | 'defaultGroupingExpansionDepth'
+  >,
+) => {
+  const groupingColDef = React.useMemo<GridColDef>(() => {
+    const propGroupingColDef = props.groupingColDef;
+
+    const baseColDef: GridColDef = {
+      ...GRID_TREE_DATA_GROUP_COL_DEF,
+      headerName: apiRef.current.getLocaleText('treeDataGroupingHeaderName'),
+    };
+    let colDefOverride: Partial<GridColDef>;
+
+    if (isFunction(propGroupingColDef)) {
+      const params: GridColDefOverrideParams = {
+        colDef: baseColDef,
+      };
+
+      colDefOverride = propGroupingColDef(params);
+    } else {
+      colDefOverride = propGroupingColDef ?? {};
+    }
+
+    return {
+      ...baseColDef,
+      ...colDefOverride,
+    };
+  }, [apiRef, props.groupingColDef]);
+
+  const updateColumnsPreProcessing = React.useCallback(() => {
+    const addGroupingColumn = (columns: GridColumns) => {
+      if (!props.treeData) {
+        return columns;
+      }
+
+      const index = columns[0].type === 'checkboxSelection' ? 1 : 0;
```

Comment:
Would checking for the `field` be more reliable?

```suggestion
      const index = columns[0].field === '__check__' ? 1 : 0;
```

Reply:
I started checking the type instead of the field, because on  the row grouping by column you can have several grouping columns and therefore the field is not static, it becomes something like `__row_group_by_column_${originalFIeld}`
So I am checking the type as it would be more reliable in this scenario.
But for the checkbox column, I suppose both are equivalent no ?

Reply:
For now both appear to yield the same result so keep as it is. However, in the future, checking the `field` seems to be better since it's unique while the type can repeat.

Reply:
If there are no columns it crashes here. I could decide to have only the grouping column.

Reply:
Fixed